clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -264,37 +276,40 @@ public boolean mkdirs(Path path, FsPermission fsPermission) throws IOException {
     public FileStatus getFileStatus(Path path) throws IOException {
         LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ getFileStatus, path: {} $$$$$$$$$$$$$$$$$$$$$$$$$$$$ ", path.toString());
         ObjectLocation objectLoc = pathToObjectLocation(path);
-        if (objectLoc == null) {
-            throw new FileNotFoundException(path.toString());
-        }
         try {
             ObjectsApi objectsApi = new ObjectsApi(this.apiClient);
             ObjectStats objectStat = objectsApi.statObject(objectLoc.getRepository(), objectLoc.getRef(), objectLoc.getPath());
-            long length = 0;
-            Long sizeBytes = objectStat.getSizeBytes();
-            if (sizeBytes != null) {
-                length = sizeBytes;
-            }
-            long modificationTime = 0;
-            Long mtime = objectStat.getMtime();
-            if (mtime != null) {
-                modificationTime = TimeUnit.SECONDS.toMillis(mtime);
-            }
-            Path filePath = path.makeQualified(this.uri, this.workingDirectory);
-            return new FileStatus(length, false, 0, 0, modificationTime, filePath);
+            return convertObjectStatsToFileStatus(objectStat);
```

Comment:
Is there a different way to declare conversion - I don't want to extent FileStatus to LakeFSFileStatus when I'm not adding anything. Also, it will not require me to convert between types of collections that I'm working on because LakeFSFileStatus is a different type?

Reply:
> Is there a different way to declare conversion - I don't want to extent FileStatus to LakeFSFileStatus when I'm not adding anything.

Makes sense, feel free to drop the comment.

Reply:
Yeah, Java generics are hard to manage with variance.  Scala allows explicit variance annotations, but it's not like a Scala iterator would solve the problem, because this method needs to return a Java iterator.

But I think we can still have an iterator of `FileStatus` objects whose `next` method happens to return (dynamic) type `LakeFSFileStatus` as a (static) type `FileStatus`.  In fact this is what I see in S3AFileSystem: `listFiles` [returns a proxy type](https://github.com/apache/hadoop/blob/359c0c1d1edb5ef3901e95f09364be6bd742866e/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L4402) returns an iterator LocatedFileStatus that actually [downcasts its S3ALocatedFileStatus](https://github.com/apache/hadoop/blob/35ca1dcb9d9b14e31ad5d0e327a556cc0529f4ce/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L4475).  But in even more fact, I think our `next` could just return the more specific LakeFSFileStatus: this is just [return type covariance](https://docs.oracle.com/javase/tutorial/java/javaOO/returnvalue.html) that happens to be part of the implementation of a generic interface.

Doesn't need to be in this PR, but I don't see why we cannot do what @talSofer suggests.